Causal non param

NathanielF opened the pull request:
# Causal inference with propensity scores and Non-parametric methods

This is a WIP show-casing causal inference methods using non-parametric models and propensity score methods to highlight the use of BART and (if i can get it to work) Dirichlet Dependent regression in their usage in causal inference. 

We will examine two data sets (1) the NHEFs data set discussed in the What-If causal inference book 
and (2) and health expenditure data set discussed in the bayesian nonparametrics book for causal inference and missing data book.  

This is related to the issue: #623 https://github.com/pymc-devs/pymc-examples/issues/623

+ [X] Notebook follows style guide https://docs.pymc.io/en/latest/contributing/jupyter_style.html
+ [X] PR description contains a link to the relevant issue:
  * a tracker one for existing notebooks (tracker issues have the "tracker id" label)
  * or a proposal one for new notebooks
+ [x] Check the notebook is not excluded from any pre-commit check: https://github.com/pymc-devs/pymc-examples/blob/main/.pre-commit-config.yaml


### Helpful links
* https://github.com/pymc-devs/pymc-examples/blob/main/CONTRIBUTING.md

drbenvincent commented:
Feel free to tag me for review once this is ready.

AlexAndorra commented:
Same about review if needed @NathanielF

juanitorduz commented:
Me too 😁 ! (Also any help with BART)

NathanielF commented:
Hi folks - @drbenvincent , @AlexAndorra , @juanitorduz 

Tagging this one as ready for review. 

The broad idea of the example was to show how to implement and assess propensity score weighting methods in a bayesian non-parametric approaches to causal inference using the flexibility of BART  to estimate the propensity scores. 

This turned out to be quite a rich vein. I cover robust and and doubly robust approaches to IPws and demonstrate their use on two data sets as "corrective" approaches. In (i) they work well and (ii) they work poorly. 

In the case where they work poorly i show how the doubly robust estimator and double ML methods can be used to correct the bias induced by model mis-specification. 

Then finally i contrast the propensity score methods in (ii) at getting at the ATE with a mediation style analysis (basically cribbed from Ben's example here) to show that the correct causal inference approach requires additional structural commitments about direction of effects. 

I'm possibly doing too much here. But both data sets cover effects of smoking, difficuties of causal inference and have a natural trajectory where i try and lead the reader to the conclusion that naive Double ML approaches (even with the corrective moves) are not sufficient to model the complexity of real-world causal inference. 

Any and all feedback welcome.

juanitorduz commented:
@NathanielF WOW! Amazing work! Really! This will bring causality and Bayes much closer (have you thought about a PyMC Con talk?, it would be great). I left some minor comments, hopefully improving the current version, which is in an excellent state already! 

Ping me whenever you want another feedback round!

NathanielF commented:
Thanks so much @juanitorduz for the review and the comments. Will try and address them this weekend. Hadn't thought about pymcon. But it's a good idea.

NathanielF commented:
<p>Good idea. Moved up </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896838648'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I added a DAG above but it's a DAG of the propensity score scenario. I wanted to keep the causal structure minimal in this example as it makes the ending more "dramatic" to show that we can't get away with just assuming strong ignorability but should be clearer about the DAG. I've fleshed out a paragraph discussing how propensity score methods are often seen as an approach to causal inference where we can get away with being non-commital about the DAG... I hope that makes sense?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896838297'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added a link to the CRC book PyMC book chapter on BART for some reason the pre-commit checks broke when i tried to link to the pymc-bart docs/repo. They wanted a cross-reference link rather than a url... </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896838468'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I've plotted all the propensity scores. Nothing too large here but i've noted that the positivity constraint is important. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896838736'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Good idea. I've pulled out the various adjustment methods so as to be able to reuse them. Makes things allot neater</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896838812'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Adjusted. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896838972'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Yeah, this has been challenging! I try to still use the full posterior downstream but it would be ideal to have in one model.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896839060'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>See comment on DAGs above. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896839163'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Done. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896839318'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
This should be good now for another pass @juanitorduz Thanks so much!

AlexAndorra commented:
<p>I agree with Juan that the DAG would make this clearer, but now I understand why you didn't include it yet Nathaniel ;)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896838297'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913713968'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913713969'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Adjusted</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913713977'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913713981'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913713985'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added a more explanatory paragraph preceding the table: </p><p> </p><blockquote>We see that there is some overall differences between the two groups, but splitting this out further we might worry that the differences are due to how the groups are imbalanced across the different covariate profiles in the treatment and control groups. That is to say, we can inspect the implied differences across different strata of our data to see that we might have imbalance across different niches of the population. This imbalance indicative of some selection effects into the treatment status is what we hope to address with propensity score modelling.</blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913713995'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913714000'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Ooops, you're right. Fixed. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913714003'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Yes. Clarified this. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913714007'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913714012'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed. Added a clarifying comment on the "clips..." sentence. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913714016'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>Ah yeah, that's better, thanks Nathaniel</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913713995'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>What if you used random draws of the ps, instead of the means? Not ideal, but that should already induce more variability in the ps.</p><p>Or maybe you can use the posterior predictive samples of treatment (modeled by the previous propensity score model) instead of the observed treatment?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896839060'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896838887'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Yes, i'll demonstrate how we can be more Bayesian in post-processing the posterior predictive distributions in a few examples below. </p><p>I'm not getting too hung up on it in this case because (a) i want to just quickly demonstrate that regression modelling can include propensity scores and they are useful here and (b) while it works well in this case it will fail below because the propensity score centered at the mean derived from the BART model is misspecified. So i can argue for more flexible approaches with better relationship to understanding the uncertainty in these estimates. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896839060'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>No, we fitted both above. Here we're plotting the idata_logit results. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922328972'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Totally fair. </p><p> </p><blockquote>There is allot going on in this plot so it's worth walking through it a bit more slowly. In the first panel we have the distribution of the propensity scores across both the treatment and control groups. In the second panel we have the raw outcome data plotted again as a distribution split between the groups. Additionally we have presented the expected values of the outcome and the ATE if it were naively estimated on the raw outcome data. Finally on the third panel we have the re-weighted outcome variable - reweighted using the inverse propensity scores, and we derive the ATE based on the adjusted outcome variable. The distinction between the ATE under the raw and adjusted outcome is what we are highlighting in this plot. </blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922328989'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>In this case it's not. But more generally the doubly robust estimator is explicitly designed to reduce the variance in the estimate. I'll be more clear about this. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329075'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Adjusted</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329050'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329065'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Adjusted</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329105'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Good catch</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329120'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I do spell in some more detail the latex formula specification of the doubly-robust estimator below. I wait until after we see that the BART model fails on the next data set. Do you think we should move it up? Add more? </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329012'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>Awesome, thanks!</p><p>Just two things:</p><ul><li>"allot"--&gt; "a lot"</li><li>I think there is a typo in the third panel: you're writing you're computing E(Y(1)) - E(Y(0)), but the value is E(Y(0)) - E(Y(1))</li></ul>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922328989'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>Still a typo: "deepter" --&gt; "deeper"</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329092'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>I see, sounds like a plan!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896839060'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I include them largely to help recover the 3.4ish estimate that is reported for <code>trt</code> in Hernan's book. Without the interaction we get less well matched numbers. </p><p>Without the propensity score we have to include the other covariates to achieve the same  </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329146'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Thanks! (: </p><p> </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329092'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I'll try and invert this everywhere. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329032'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>Yeah I think that'd be better to introduce it here, otherwise the reader is left wondering.</p><p>Really understanding the two steps of PS weighting, and how the different methods impact the ATE seems crucial to this NB</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329012'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra reviewed (changes requested):
Finished reviewing the whole NB @NathanielF ! Well done 👏 
I left a bunch of comments. LMK if anything is unclear

NathanielF commented:
<p>Softened the tone here and emphasised that we'd be exploring this result as it appears striking given our EDA plots, that the average could be shifted more than a log-dollar unit e.g. 6 - 7.4.....</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925458837'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added a line to clarify our post-processing efforts</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925458841'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I don't know at this point, but i'm re-worked the discussion a little to lean on our EDA a bit more at this point. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925458845'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Thanks. I moved this up too. I placed it beneath the code definitions. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925458851'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Adjusted</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921458'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Nice!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921481'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Adjusted. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921523'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Yes, exactly. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921548'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Changed. Also increased target accept. No divergences now. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921572'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I added a (hopefully clarifying) paragraph</p><p> </p><blockquote>These equations are subtle, and the potential outcomes notation is perhaps a little obscure, but it’s the notation we inherit. The key point is just that in NDE the treatment regime for the mediation effects are fixed and the values for the outcome varied. In the NIE the values for the mediation effects are varied and the treatment regime for the outcome variable is fixed.</blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921598'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Moved up.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921632'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added result to the kwargs. It's the idata object from the original fit of the mediation model. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921656'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>See above. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921678'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>changed this. We're using the observed mediation values here in any case. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921707'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed. Updated with slightly stronger conclusion</p><p> </p><blockquote>This is just how inference is done. You encode your knowledge of the world and update your views as evidence accrues. Causal inference requires commitment to a structural view of the world and we cannot pretend to ignorance when it makes inference indefensible.</blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921757'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>I think i've softened the language enough to not presuppose that the reader agrees that the naive is estimate is wrong. Instead i've suggested it's wrong and we'll explore the question as we go. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921504'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Changed this everywhere i think. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925458834'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<blockquote>This is a&nbsp;<strong>suspicious</strong>&nbsp;result. Evaluated at the expected values of the posterior propensity score distribution the robust IPW estimator of ATE suggests a substantial difference in the treatment and control groups. When a simple difference in the averages of the raw outcome show close to 0 differences but the re-weighted difference shows a movement of more than 1 on the log-scale. What is going on? In what follows we’ll interrogate this question from a couple of different perspectives but you should be pretty dubious of this result as it stands.</blockquote><blockquote>What happens if we look at the posterior ATE distributions under different estimators?</blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925921504'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<blockquote>This is a&nbsp;<strong>suspicious</strong>&nbsp;result. Evaluated at the expected values of the posterior propensity score distribution the robust IPW estimator of ATE suggests a substantial difference in the treatment and control groups. When a simple difference in the averages of the raw outcome show close to 0 differences but the re-weighted difference shows a movement of more than 1 on the log-scale. What is going on? In what follows we’ll interrogate this question from a couple of different perspectives but you should be pretty dubious of this result as it stands.</blockquote><blockquote>What happens if we look at the posterior ATE distributions under different estimators?</blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925458837'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925458831'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Clarified a little. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1925458828'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329032'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Done. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922329012'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Changed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1922328989'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>changed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1913713988'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>There are definitely extreme values in the propensity scores. This increases as you add more depth to the tree worsening the discrepancy. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1896839248'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
> Finished reviewing the whole NB @NathanielF ! Well done 👏
I left a bunch of comments. LMK if anything is unclear

Thanks so much for you detailed review! I think it's much stronger now for your suggestions! 
I think i've addressed substantially all the main questions or pushes you had now.

NathanielF commented:
<p>Added a sentence </p><p> </p><blockquote>A propensity scores is the probability of receiving a treatment status (e.g. smoker/non-smoker, divorced/married) for each individual in the population under study. An individual’s propensity-for-treatment helps analyse the impact of each such treatment status over and above the other available treatment options. In this notebook.... </blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782296'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added link and fixed typo. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782316'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<blockquote>We will use the&nbsp;<em>potential outcomes</em>&nbsp;notation to discuss causal inference - we draw extensively on work in&nbsp;Aronow and Miller [<a href="https://pymcio--622.org.readthedocs.build/projects/examples/en/622/causal_inference/bayesian_nonparametric_causal.html#id14" target="_blank">2019</a>], which can be consulted for more details. But very briefly, when we want to discuss the causal impact of a treatment status on an outcome&nbsp;Y we will denote the outcome&nbsp;Y(t) as the outcome measure&nbsp;Y under the treatment regime&nbsp;T=t</blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782336'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Done. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782362'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>It's all from the Aronow and Miller book referenced in the previous chapter. Also discussed in the Bayesian non-parametric book too. Added that it's also sometimes called exchangeability. Made the debt to Aronow and Miller clearer above. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782392'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782419'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Pretty much yes. Emphasised that above</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782441'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added a visual of the covariate stratified across the propensity score to underline the differences across treatment groups under various covariate profiles.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782462'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Hidden</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782508'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Hidden</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782532'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Hidden</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782554'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Added</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782575'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Left this one as is... already have EDA here. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782599'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782627'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Changed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782647'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Hidden</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782678'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782698'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782724'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Separated and hid the plotting code</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782739'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Made the font bigger</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782763'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Hidden</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782794'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
I think this is good now @AlexAndorra ,@drbenvincent , @juanitorduz 

Let me know of any concerns, but ideally @AlexAndorra I'd like to get this merged before March 1st.

AlexAndorra commented:
Thanks @NathanielF ! Will review ASAP. Definitely before March 1st, so that it's ready for the webinar

aloctavodia commented:
<p>agreed</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782488'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>added another plot</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1926782599'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>This is really good. I've re-structured this paragraph with these comments in mind! Thanks</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1946148721'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1946148759'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Reworked this too</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1946148785'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
Thanks for approving @aloctavodia !

AlexAndorra reviewed (approved):
Just added two comments, but looks good now, thanks a lot @NathanielF !

NathanielF commented:
<p>Clarified this a bit</p><p>The balancing tests we've just seen are to show balance of covariate conditional on the propensity scores. With good demonstrated balance i.e mean differences close to zero, suggests that strong ignorability holds and then we can plausibly use the propensity scores</p><p>: </p><blockquote>In an ideal world we would have perfect balance across the treatment groups for each of the covariates, but even approximate balance as we see here is useful. When we have good covariate balance (conditional on the propensity scores) we can then use propensity scores in weighting schemes with models of statistical summaries so as to "correct" the representation of covariate profiles across both groups.</blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1962560493'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>It is very close in approach to the re-weighting strategies use by pollsters! I started looking into the strata re-weighting stuff after looking into MrP stuff here:<a href="https://bambinos.github.io/bambi/notebooks/mister_p.html" target="_blank">https://bambinos.github.io/bambi/notebooks/mister_p.html</a></p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1962560493'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
<p>Yeah, that was a mess. </p><p>Rewrote: </p><blockquote>It's difficult to see a clear pattern in this visual. In both treatment groups, when there is some significant sample size, we see a mean difference close to zero for both groups.</blockquote>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/622/discussion/#comment-1962560516'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

NathanielF commented:
Addressed those comments @AlexAndorra . I'm happy if you're happy. Feel free to merge! 

Thanks again @AlexAndorra , @drbenvincent , @juanitorduz and @aloctavodia . The piece is much stronger for all your feedback!

AlexAndorra commented:
Awesome @NathanielF ! Can we merge without the pre-commit build passing?

NathanielF commented:
I think so. That check was never used before and @maresb opened an issue about it here: https://github.com/pymc-devs/pymc-examples/issues/638

maresb commented:
Ya, that one pre-commit check is passing, and that's what counts.

We just need someone with admin to disable the broken failing check as per #638.

AlexAndorra commented:
Ok, merging then 🍾

NathanielF commented:
Thanks so much @AlexAndorra !
